### PR TITLE
Fixed modded overworld biomes not spawning [1.16.x]

### PR DESCRIPTION
**Pull request by ghost**
Adapted the same patch on 1.15.2 to 1.16.1 Forge to allow modded forge biomes from `BiomeManager` to be used in `BiomeLayer` with weighted lists. Should be straight forward and crucial patch, of course let me know if there are changes required to the styling or similar.

PR will allow modded biomes to spawn in the overworld.  Test mod `modded_overworld_biomes_test` included, adding the biome `red_biome` with bright red grass so it can be found easily.

### CODE REVIEW

**Review by marchermans — COMMENTED** [review]
In general this looks good.
However due to the nature of the biome and worldgen systems in 1.16, I am not sure this will be pulled into the main branch before those systems stabilize. 

This is not a rejection of the pr, just a heads up that it might take a bit and that tu ou might need to keep it up to date for a bit while other changes are merged. Up until the point either we think it can be merged or untill mojang decides that the biome and worldgen systems is ready and stable. 

Greets orion

**Comment by marchermans** [comment]
Additionally things like proper documentation of why this is needed, and test mods that show the system in action are missing.

@Xerousso

**Comment by ghost** [comment]
Yeah, I'll see what I can do about some test mods, I was just reading the contribution guide and it said for new hooks test mods are required, which this isn't as the hooks already exist in `BiomeManager`, this patch is just to make them actually work.

The PR is required as currently the information given to `BiomeManager` by adding biomes and similar is not actually taken into consideration for the `BiomeLayer` when creating the world, meaning that right now people can only add biomes to the overworld by making their own world type and layers, which is rather poor, or achieve it through coremodding which as we all know is extremely incompatible.

**Comment by Xerousso** [comment]
Also I was told that the failure of the jenkins build was on jenkins end, and I'm hoping that it won't negatively impact my PR by making people think that it doesn't build correctly?  If anyone has the ability to rerun that it would be appreciated :)

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (pupnewfster) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Is there a reason these fields are being deleted rather than just no longer using them?

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (pupnewfster) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: There is no reason to be changing the formatting of vanilla classes (such as moving the bracket to the next line). And the formatting should match the rest of the file, which means that for patches brackets go on the same line not on the next one (there are a lot of cases below where you are putting them on the next one).

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (pupnewfster) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: I am not actually positive on the stance of Forge Registry vs Vanilla Registry use in patches, but is there a reason this is using the vanilla registry instead of the Forge Registry for biomes?

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: As I said this is an adaption of a previous patch that was present in 1.15.2, I can add them back in if you would like, although they do have no use.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (Xerousso) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch:49]: There is not a reason, again it was an adaptation.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Ah thanks, this is my first time contributing so I was not sure.

**Comment by LexManos** [comment]
```> Task :forge:applyPatches FAILED

  Apply Patch: net/minecraft/world/gen/layer/BiomeLayer.java.patch

    Hunk #1 Failed: Cannot find hunk target

    Hunk #2 Failed: Cannot find hunk target```

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Looking at it closer, it appears that the forge registries do not have an option to get the numerical id of a biome, correct me if I'm wrong.  This is why vanilla registries are required here.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Fixed

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (Shadows-of-Fire) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: They... do, but that requires a cast to the actual underlying type (ForgeRegistry) rather than the field type of IForgeRegistry

Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Ah, as such?:
```java
return ((ForgeRegistry) ForgeRegistries.BIOMES).getID(getWeightedBiomeEntry(net.minecraftforge.common.BiomeManager.BiomeType.ICY, context).biome);
```
I suppose I should replace it with this?  Let me know what you think

Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: > and I do not prefer the ForgeRegistry cast hack
> as it ties things to one implementation
> - tterrag

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Re-added

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (valoeghese) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Not sure if extra whitespace in a patch is a problem or not, since forge prefers patches to be as small as possible. But it's probably fine.

**Review by valoeghese — APPROVED** [review]

**Comment by ghost** [comment]
By the way, if the PR is held back because things change in 1.16.2, feel free to hit me up to port my PRs, I'll be more than happy to do so :)

**Review by TelepathicGrunt — CHANGES_REQUESTED** [review]
Looks good! The only thing I would change is line 72 and remove the `context.func_202696_a(totalWeight / 10) * 10` as vanilla biomes are already multiples of 10 which makes the isTypeListModded check useless. So that line can just be `int weight = context.func_202696_a(totalWeight)` for a slightly smaller patch and 1 less check.

**Review by TelepathicGrunt — APPROVED** [review]

**Comment by ghost** [comment]
Lmao what?  I change the title, it rechecks and this time fails?

**Comment by srs-bsns** [comment]
> Looks good! The only thing I would change is line 72 and remove the `context.func_202696_a(totalWeight / 10) * 10` as vanilla biomes are already multiples of 10 which makes the isTypeListModded check useless. So that line can just be `int weight = context.func_202696_a(totalWeight)` for a slightly smaller patch and 1 less check.

The vanilla biome weights are multiples of 10 because Forge adds them in the BiomeManager to have a weight that is 10x their normal incidence so that mods can add biomes to be much rarer (weight < 10). The call to `func_202696_a` (`LazyAreaLayerContext#random`), when there are no mod-added biomes present, reduces the total weight value so that the call to `#random` properly mimics vanilla behavior.

**Review by Mysterious-Dev — APPROVED** [review]

**Comment by ghost** [comment]
> 
> 
> > Looks good! The only thing I would change is line 72 and remove the `context.func_202696_a(totalWeight / 10) * 10` as vanilla biomes are already multiples of 10 which makes the isTypeListModded check useless. So that line can just be `int weight = context.func_202696_a(totalWeight)` for a slightly smaller patch and 1 less check.
> 
> The vanilla biome weights are multiples of 10 because Forge adds them in the BiomeManager to have a weight that is 10x their normal incidence so that mods can add biomes to be much rarer (weight < 10). The call to `func_202696_a` (`LazyAreaLayerContext#random`), when there are no mod-added biomes present, reduces the total weight value so that the call to `#random` properly mimics vanilla behavior.

How will reducing all weights by the same factor affect how `#random` does vanilla behavior?

**Comment by srs-bsns** [comment]
> How will reducing all weights by the same factor affect how `#random` does vanilla behavior?

When calculating a pseudorandom number with a provided bound, the random number that is returned is the next pseudorandom number in the squence that is put through a modulus operation.
So, in this case, for example: `(12345678 % 6) * 10` results in `0` and `12345678 % 60` results in `18`, which would lead to the return of different items from the list. Thus, dividing the total weight by 10 before getting the next random, (in the case that the biome list is unmodified,) preserves the vanilla selection of biome. Without doing this, any given world, even when there are no added biomes, would have a drastically different biome layout than a vanilla world using the same seed. (This was likely @Adubbz intention, and fallows the Forge axiom of preserving vanilla behaviour.)

As I commented in your other PR (https://github.com/MinecraftForge/MinecraftForge/pull/7000#issuecomment-662168707) it would probably be prudent to wait until 1.16.2 has been released in a couple of weeks, since this code will likely become vestigial.

**Comment by TelepathicGrunt** [comment]
Yeah the suggestion I made on the random was my mistake but as for the prudent part, that's really unfounded concerns as BiomeLayer remains untouched in 1.16.2. This is a 1.15.2 hook that can be re-implemented into 1.16.1 and 1.16.2 with no change and without this hook, many biome mods are unable to port at all making this a high impact issue. 

It is highly unlikely that Mojang will change biomelayers between the current snapshot and the 1.16.2 full release. If for some god-forsaken reason they do change it, we can make another pr to re-implement the hook in a new form without burdening the forge maintainers trying to update Forge. Plus, we don't even know when 1.16.2 could come out. Some people say a week. others say two weeks. Maybe more. We simply don't know and my take is that trying to predict what Mojang is doing is not very wise and most likely isn't going to turn out right.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (marchermans) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: You need to add the Forge method markers. Additionally please check that this is at the end of the file.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (marchermans) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Remove the not needed whiteline.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (marchermans) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: You could probably get rid of this line as well.
Just use `type.ordinal()` in the couple of places you are using it.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (marchermans) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch:4]: Are these original fields still in use?
If not: Put a  `@Deprecated` in front of them (on the same line) and a comment behind them referencing their replacement (also on the same line)

Eventhough these fields might be private, if vanilla code becomes unused by forge changes. We should keep track of this. Not just for people using the code, but also for porters who port the patch in the future. It is then always good to know what happens with a field.

Do not remove the field since that can also cause issue. As I said a `@Deprecated` and a comment should suffice.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (marchermans) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Please remove this whiteline.
It is not needed.

**Review by marchermans — APPROVED** [review]
This is starting to look good.
@Xerousso Could you check the Randomization issue that @TelepathicGrunt mentioned, and make sure that, if no modded biomelayer is registered, the behaviour is identical to vanilla?

@Xerousso Additionally it would be amazing if you updated your initial PR opening message. We are trying to get our documentation to a better place. And as such it would be nice if you could provide a minimal example (a minimal amount of line suffices) that shows the usage, or reference a test mod, if it exists. If no testmod exists, please additionally create one.

Greets,
Orion

P.S.: We discussed this, since it does not look like that this will be touched by the 1.16.2 changes, we will probably pass this on for merging as soon as you did the cleanup and documentation.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: This is at the end of the file, what are Forge method markers @OrionDevelopment ?

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Sure, will do

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch:4]: Sure, will do

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Sure, will do

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: I am not quite sure as to what you mean @OrionDevelopment ?  Which line?

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (marchermans) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Look at other patched vanilla classes which have methods that have been added by forge.
There are two comments: ///////// FORGE START /////////// and /////////// FORGE END //////////// or something in that direction.
These enclose the methods that forge added to that class, and allow us to easily identify the methods which are not from the vanilla code base and are very handy during porting.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (marchermans) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: See this line 14 of your patch file?. It defines the variable idx as type.ordinal().
And is then only used todo look ups 3 times.
This variable as such is not really needed, and can be removed to reduce the patch size.

Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Ah, will remove

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Ah, makes sense.  Done.

**Comment by ghost** [comment]
Still need to find a place to learn how to make a test mod, I was told to look at other PRs but I haven't found anything but patches.

**Comment by srs-bsns** [comment]
> that's really unfounded concerns as BiomeLayer remains untouched in 1.16.2.

My opinion was more than just about this patch, but whatever, it's not that important.

With respect to this patch, `#getWeightedBiomeEntry` probably no longer needs to be here. The method was patched into `GenLayerBiome` back in 1.7.10 because it required access to the protected superclass method `GenLayer#nextInt`. This is not the case anymore, since in 1.14 (or maybe it was 1.13, but I don't have any old artifacts kicking around to check.) RNG is handled by an external `LazyAreaLayerContext` object that can simply be passed to an external method call.

I suggest that `getWeightedBiomeEntry` be moved to the Forge BiomeManager where it belongs, and external calls be made to it to select a biome at random. This will reduce the overall patch size, and any changes that need to be made to this process in the future should be exclusive to Forge classes.

**Comment by ghost** [comment]
https://gist.github.com/Xerousso/e135fd69efa617444a11ac9e9cdfc2b9
Would adding it to `BiomeManager` as seen in this gist be preferable?

**Comment by srs-bsns** [comment]
> 
> 
> https://gist.github.com/Xerousso/e135fd69efa617444a11ac9e9cdfc2b9
> Would adding it to `BiomeManager` as seen in this gist be preferable?

Basically that, yes.

**Review by ilja615 — APPROVED** [review]
I approve this PR and I hope this PR will be added to forge so that we can make ModdedBiome, I am looking forward to creating my own ModdedBiomes in 1.16.1 as well as to see what amazing creations the other modders would  make!
- Greetings, 
Ilja 615

**Review by CorgiTaco — APPROVED** [review]
I approve This PR is rather essential because custom world types in 1.16 have been rather defective and world types for small biome adding mods isn't necessarily efficient.

File: src/main/java/net/minecraftforge/common/BiomeManager.java
Comment (srs-bsns) [inline src/main/java/net/minecraftforge/common/BiomeManager.java:166]: This method can probably be renamed to `getWeightedBiome` and just return the `Biome`, since the weight value is only needed here and not in BiomeLayer.

Comment (ghost) [reply src/main/java/net/minecraftforge/common/BiomeManager.java:166]: True, but it may be useful for other things that we haven't thought about to use this method and get the weight

Comment (srs-bsns) [reply src/main/java/net/minecraftforge/common/BiomeManager.java:166]: Even if the weight value is useful for other processes, those other processes should probably be handled within `BiomeManager` or in other processes called from it, and not done within `BiomeLayer`, which would also require further patching it. At the point that this method returns, it should only be the `Biome`, since that it all that is consequential in `BiomeLayer`.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (marchermans) [inline patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: Please add a comment stating why they are deprecated and what their replacement is, if any.

File: src/main/java/net/minecraftforge/common/BiomeManager.java
Comment (marchermans) [reply src/main/java/net/minecraftforge/common/BiomeManager.java:166]: @Xerousso You can leave it like this.
It does not hurt to have it return `BiomeEntry` instead of biome.

File: src/test/java/net/minecraftforge/debug/world/ModdedOverworldBiomesTest.java
Comment (marchermans) [inline src/test/java/net/minecraftforge/debug/world/ModdedOverworldBiomesTest.java]: See the other mods on how todo these. Most notably missing is a way to disable a given testmod.

**Review by marchermans — CHANGES_REQUESTED** [review]
Some small nitpickings.

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (Xerousso) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch:18]: @OrionDevelopment Should I do a javadoc for each one or just a standard comment?

File: src/test/java/net/minecraftforge/debug/world/ModdedOverworldBiomesTest.java
Comment (Xerousso) [reply src/test/java/net/minecraftforge/debug/world/ModdedOverworldBiomesTest.java:50]: I've checked 10 different test mods and have not found a way to disable a given one in their code?

File: patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch
Comment (ghost) [reply patches/minecraft/net/minecraft/world/gen/layer/BiomeLayer.java.patch]: LexManos and diesieben07 said its pointless to do a comment or deprecate these at all as they are private fields.

File: src/test/java/net/minecraftforge/debug/world/ModdedOverworldBiomesTest.java
Comment (ghost) [reply src/test/java/net/minecraftforge/debug/world/ModdedOverworldBiomesTest.java]: Done.

**Comment by marchermans** [comment]
@Xerousso Why did you remove the  `@Deprecated` markers?
`Has been resolved on Discord`Sio disregard

**Review by marchermans — APPROVED** [review]
Readup on the discord discussion. Sorry missed that.
Okey this PR looks good in my eyes.

Assigning it to lex.

**Pull request merged**
